Error while checking build/O2/o2-dataflow-cs8 for 59697c0e2875afdaccb712651dd990ec91362e40 at 2022-10-01 04:23:
```
## sw/BUILD/O2-latest/log
100% tests passed, 0 tests failed out of 434
 92/106 Test #106: test_Framework_test_SingleDataSource ....................***Failed    3.04 sec
[ERROR] pid 25563 (internal-dpl-injected-dummy-sink) crashed with 1
[ERROR] SEVERE: Device internal-dpl-injected-dummy-sink (25563) returned with 1
99% tests passed, 1 tests failed out of 102
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/9881/59697c0e2875afdaccb712651dd990ec91362e40/build_O2_o2-dataflow-cs8/pretty.html).

--- a/Detectors/MUON/MCH/Raw/Decoder/include/MCHRawDecoder/DataDecoder.h
+++ b/Detectors/MUON/MCH/Raw/Decoder/include/MCHRawDecoder/DataDecoder.h
@@ -177,7 +177,6 @@ class DataDecoder
   /// Compute the time of all the digits that have been decoded in the current TimeFrame
   void computeDigitsTimeBCRst();
   void computeDigitsTime();
-  void checkDigitsTime();
 
   /// Get the vector of digits that have been decoded in the current TimeFrame
   const RawDigitVector& getDigits() const { return mDigits; }

--- a/Detectors/MUON/MCH/Raw/Decoder/include/MCHRawDecoder/ErrorCodes.h
+++ b/Detectors/MUON/MCH/Raw/Decoder/include/MCHRawDecoder/ErrorCodes.h
@@ -25,16 +25,17 @@ enum ErrorCodes {
   ErrorParity = 1,                           // 1
   ErrorHammingCorrectable = 1 << 1,          // 2
   ErrorHammingUncorrectable = 1 << 2,        // 4
-  ErrorBadClusterSize = 1 << 3,              // 8
-  ErrorBadPacketType = 1 << 4,               // 16
-  ErrorBadHeartBeatPacket = 1 << 5,          // 32
-  ErrorBadIncompleteWord = 1 << 6,           // 64
-  ErrorTruncatedData = 1 << 7,               // 128
-  ErrorBadELinkID = 1 << 8,                  // 256
-  ErrorBadLinkID = 1 << 9,                   // 512
-  ErrorUnknownLinkID = 1 << 10,              // 1024
-  ErrorInvalidDigitTime = 1 << 11,           // 2048
-  ErrorNonRecoverableDecodingError = 1 << 12 // 4096
+  ErrorBadSyncPacket = 1 << 3,               // 8
+  ErrorBadHeartBeatPacket = 1 << 4,          // 16
+  ErrorBadDataPacket = 1 << 5,               // 32
+  ErrorBadClusterSize = 1 << 6,              // 64
+  ErrorBadIncompleteWord = 1 << 7,           // 128
+  ErrorTruncatedData = 1 << 8,               // 256
+  ErrorBadELinkID = 1 << 9,                  // 512
+  ErrorBadLinkID = 1 << 10,                  // 1024
+  ErrorUnknownLinkID = 1 << 11,              // 2048
+  ErrorBadHBTime = 1 << 12,                  // 4096
+  ErrorNonRecoverableDecodingError = 1 << 13 // 8192
 };
 
 uint32_t getErrorCodesSize();

--- a/Detectors/MUON/MCH/Raw/Decoder/src/DataDecoder.cxx
+++ b/Detectors/MUON/MCH/Raw/Decoder/src/DataDecoder.cxx
@@ -814,21 +814,6 @@ void DataDecoder::computeDigitsTimeBCRst()
 
 //_________________________________________________________________________________________________
 
-void DataDecoder::checkDigitsTime()
-{
-  for (auto& digit : mDigits) {
-    auto& d = digit.digit;
-    auto& info = digit.info;
-    auto tfTime = d.getTime();
-    if (tfTime == DataDecoder::tfTimeInvalid) {
-      // add invalid digit time error
-      mErrors.emplace_back(o2::mch::DecoderError(info.solar, info.ds, info.chip, ErrorInvalidDigitTime));
-    }
-  }
-}
-
-//_________________________________________________________________________________________________
-
 void DataDecoder::computeDigitsTime()
 {
   switch (mTimeRecoMode) {

--- a/Detectors/MUON/MCH/Raw/Decoder/src/ErrorCodes.cxx
+++ b/Detectors/MUON/MCH/Raw/Decoder/src/ErrorCodes.cxx
@@ -20,7 +20,7 @@ namespace raw
 
 uint32_t getErrorCodesSize()
 {
-  return 13;
+  return 14;
 }
 
 void append(const char* msg, std::string& to)
@@ -49,11 +49,14 @@ std::string errorCodeAsString(uint32_t ec)
   if (ec & ErrorBadClusterSize) {
     append("Cluster Size", msg);
   }
-  if (ec & ErrorBadPacketType) {
-    append("Bad Packet Type", msg);
+  if (ec & ErrorBadSyncPacket) {
+    append("Bad Sync Packet", msg);
   }
   if (ec & ErrorBadHeartBeatPacket) {
     append("Bad HB Packet", msg);
+  }
+  if (ec & ErrorBadDataPacket) {
+    append("Bad Data Packet", msg);
   }
   if (ec & ErrorBadIncompleteWord) {
     append("Bad Incomplete Word", msg);
@@ -70,8 +73,8 @@ std::string errorCodeAsString(uint32_t ec)
   if (ec & ErrorUnknownLinkID) {
     append("Unknown Link ID", msg);
   }
-  if (ec & ErrorInvalidDigitTime) {
-    append("Invalid Digit Time", msg);
+  if (ec & ErrorBadHBTime) {
+    append("Bad HB Time", msg);
   }
   if (ec & ErrorNonRecoverableDecodingError) {
     append("Non Recoverable", msg);

--- a/Detectors/MUON/MCH/Raw/Decoder/src/UserLogicElinkDecoder.cxx
+++ b/Detectors/MUON/MCH/Raw/Decoder/src/UserLogicElinkDecoder.cxx
@@ -38,4 +38,42 @@ void UserLogicElinkDecoder<ChargeSumMode>::prepareAndSendCluster()
   mSamples.clear();
 }
 
+template <>
+bool UserLogicElinkDecoder<SampleMode>::checkDataHeader()
+{
+  int chipAddMin = mDsId.elinkIndexInGroup() * 2;
+  int chipAddMax = chipAddMin + 1;
+
+  // the chip address from the SAMPA header must be consistent with
+  // the e-Link index
+  int chipAdd = mSampaHeader.chipAddress();
+  if (chipAdd < chipAddMin || chipAdd > chipAddMax) {
+    return false;
+  }
+
+  return true;
+}
+
+template <>
+bool UserLogicElinkDecoder<ChargeSumMode>::checkDataHeader()
+{
+  int chipAddMin = mDsId.elinkIndexInGroup() * 2;
+  int chipAddMax = chipAddMin + 1;
+
+  // the chip address from the SAMPA header must be consistent with
+  // the e-Link index
+  int chipAdd = mSampaHeader.chipAddress();
+  if (chipAdd < chipAddMin || chipAdd > chipAddMax) {
+    return false;
+  }
+
+  // in cluster sum mode the number of 10-bit words must be a multiple of 4
+  int nof10BitWords = mSampaHeader.nof10BitWords();
+  if ((nof10BitWords % 4) != 0) {
+    return false;
+  }
+
+  return true;
+}
+
 } // namespace o2::mch::raw

--- a/Detectors/MUON/MCH/Raw/Decoder/src/UserLogicElinkDecoder.h
+++ b/Detectors/MUON/MCH/Raw/Decoder/src/UserLogicElinkDecoder.h
@@ -73,6 +73,7 @@ class UserLogicElinkDecoder
   void completeHeader();
   void oneLess10BitWord();
   void prepareAndSendCluster();
+  bool checkDataHeader();
   void sendCluster(const SampaCluster& sc) const;
   void sendHBPacket();
   void sendError(int8_t chip, uint32_t error) const;
@@ -185,7 +186,11 @@ bool UserLogicElinkDecoder<CHARGESUM>::append10(uint10_t data10)
       setHeaderPart(data10);
       if (isHeaderComplete()) {
         completeHeader();
-        if (isSync(mSampaHeader.uint64())) {
+        if (mSampaHeader.packetType() == SampaPacketType::Sync) {
+          if (!isSync(mSampaHeader.uint64())) {
+            mErrorMessage = "badly formatted Sync packet";
+            sendError(static_cast<int8_t>(mSampaHeader.chipAddress()), static_cast<uint32_t>(ErrorBadSyncPacket));
+          }
           reset();
         } else if (mSampaHeader.packetType() == SampaPacketType::HeartBeat) {
           if (mSampaHeader.isHeartbeat()) {
@@ -198,9 +203,15 @@ bool UserLogicElinkDecoder<CHARGESUM>::append10(uint10_t data10)
             reset();
           }
         } else {
-          if (mSampaHeader.nof10BitWords() > 2) {
-            transition(State::WaitingSize);
+          if (checkDataHeader()) {
+            if (mSampaHeader.nof10BitWords() > 2) {
+              transition(State::WaitingSize);
+            } else {
+              reset();
+            }
           } else {
+            mErrorMessage = "badly formatted Data packet";
+            sendError(static_cast<int8_t>(mSampaHeader.chipAddress()), static_cast<uint32_t>(ErrorBadDataPacket));
             reset();
           }
         }

--- a/Detectors/MUON/MCH/Workflow/src/DataDecoderSpec.cxx
+++ b/Detectors/MUON/MCH/Workflow/src/DataDecoderSpec.cxx
@@ -243,7 +243,6 @@ class DataDecoderTask
       }
     }
     mDecoder->computeDigitsTime();
-    mDecoder->checkDigitsTime();
     auto tEnd = std::chrono::high_resolution_clock::now();
     mTimeDecoding += tEnd - tStart;
 

--- a/Detectors/MUON/MCH/Workflow/src/cru-page-reader-workflow.cxx
+++ b/Detectors/MUON/MCH/Workflow/src/cru-page-reader-workflow.cxx
@@ -320,9 +320,10 @@ class FileReaderTask
     }
 
     if (mPrint) {
-      std::cout << "Sending TF" << std::endl
+      std::cout << "Sending TF " << orbitMin << " (previous " << mLastTForbit << "  delta " << orbitMin-mLastTForbit << ")" << std::endl
                 << std::endl;
     }
+    mLastTForbit = orbitMin;
     auto freefct = [](void* data, void* /*hint*/) { free(data); };
     pc.outputs().adoptChunk(Output{"RDT", "RAWDATA"}, outBuf, outSize, freefct, nullptr);
 
@@ -448,7 +449,7 @@ class FileReaderTask
       // increment the total buffer size
       mTimeFrameSizes[feeID][linkID] += pageSize;
 
-      if ((triggerType & 0x800) != 0 && stopBit == 0 && pageCounter == 0 && bc == 0) {
+      if ((triggerType & 0x800) != 0 && /*stopBit == 0 && pageCounter == 0 &&*/ bc == 0) {
         // This is the start of a new TimeFrame, so we need to push a new empty TimeFrame in the queue
         if (mPrint) {
           std::cout << "tfQueue.size(): " << tfQueue.size() << std::endl;
@@ -615,6 +616,7 @@ class FileReaderTask
   bool mFullTF;               ///< send full time frames
   bool mSaveTF;               ///< save individual time frames to file
   int mOverlap;               ///< overlap between contiguous TimeFrames
+  int mLastTForbit{ 0 };
   bool mPrint = false;        ///< print debug messages
   o2::dataformats::TFIDInfo mTFIDInfo{}; // struct to modify output headers
 

